File: examples/node/roundtrip.mjs
#!/usr/bin/env node
// Small CLI integration example for a UI-automation action prior.
// The domain adapter owns the screen/action vocabulary; lineprior only sees JSONL.

import { execFileSync } from "node:child_process";
import { mkdtempSync, readFileSync, rmSync } from "node:fs";
import { tmpdir } from "node:os";
import { join, dirname } from "node:path";
import { fileURLToPath } from "node:url";

const root = join(dirname(fileURLToPath(import.meta.url)), "..");
const input = join(root, "ui_automation.jsonl");
const binary = process.env.LINEPRIOR_BIN ?? "lineprior";
const directory = mkdtempSync(join(tmpdir(), "lineprior-node-"));
const first = join(directory, "prior-1.jsonl");
const second = join(directory, "prior-2.jsonl");

try {
  const run = (...args) => execFileSync(binary, args, { encoding: "utf8", stdio: ["ignore", "pipe", "pipe"] });
  run("build", input, "--out", first);
  run("build", input, "--out", second);
  if (readFileSync(first, "utf8") !== readFileSync(second, "utf8")) {
    throw new Error("repeated builds were not deterministic");
  }
  const output = run("query", first, "--state", "cart-empty", "--top-k", "1");
  const row = output.trim().split("\n").filter(Boolean).map(JSON.parse)[0];
  if (row?.action !== "click:add-to-cart") {
    throw new Error(`unexpected query result: ${JSON.stringify(row)}`);
  }
  console.log(JSON.stringify(row));
} finally {
  rmSync(directory, { recursive: true, force: true });
}

File: examples/python/roundtrip.py
#!/usr/bin/env python3
"""Small CLI integration example for a UI-automation action prior.

The domain adapter owns the screen/action vocabulary; lineprior only sees
JSONL.  Set LINEPRIOR_BIN when the executable is not on PATH.
"""

import json
import os
import pathlib
import shutil
import subprocess
import sys
import tempfile


ROOT = pathlib.Path(__file__).resolve().parents[2]
INPUT = ROOT / "ui_automation.jsonl"
STATE = "cart-empty"


def run(*args: str) -> subprocess.CompletedProcess[str]:
    binary = os.environ.get("LINEPRIOR_BIN", "lineprior")
    return subprocess.run([binary, *args], check=True, text=True, capture_output=True)


def main() -> int:
    binary = os.environ.get("LINEPRIOR_BIN", "lineprior")
    if shutil.which(binary) is None and not pathlib.Path(binary).exists():
        print(f"lineprior executable not found: {binary}", file=sys.stderr)
        return 2

    with tempfile.TemporaryDirectory(prefix="lineprior-python-") as directory:
        first = pathlib.Path(directory) / "prior-1.jsonl"
        second = pathlib.Path(directory) / "prior-2.jsonl"
        run("build", str(INPUT), "--out", str(first))
        run("build", str(INPUT), "--out", str(second))
        if first.read_bytes() != second.read_bytes():
            raise RuntimeError("repeated builds were not deterministic")

        result = run("query", str(first), "--state", STATE, "--top-k", "1")
        rows = [json.loads(line) for line in result.stdout.splitlines() if line.strip()]
        if not rows or rows[0]["action"] != "click:add-to-cart":
            raise RuntimeError(f"unexpected query result: {rows!r}")
        print(json.dumps(rows[0], sort_keys=True))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
